### PR TITLE
New CVars: `mp_weapondrop` and `mp_ammodrop`

### LINKED ISSUE #520: Problem with ammo on droped weapons
I set on my servers that every weapon will get full ammo on buy or on spawn with thw cvar mp_refill_bpammo_weapons 2, but if someone buys a weapon, lets say a ak-47 he will have 30/90 bullets and he drop it and i pick it up, i will only have 30/0 and if i try to buy ammo it will say that its now allowed to buy, can be add to fill ammo on dropped weapons too? Or at least allow to buy ammo for dropped weapons, because it does not work, if needed i can record a vídeo and send it too

### PR DESCRIPTION
**Pull request by dystopm**
- Added cvar `mp_weapondrop` which controls behaviour on weapon drop after death. You can choose between (0) do not drop anything after death, (1) dropping your heaviest gun (default behaviour) or (2) dropping your active weapon. It works independently of mp_nadedrops cvar (and it's also an extension of it)
- Added cvar `mp_ammodrop` which controls behaviour of ammo packing on weapon boxes drop. You can choose between (0) keep ammo on player always, (1) drop ammo on weaponbox only after death and (2) drop ammo on every weaponbox you drop either death or manual drop. 
- Fix: Ammo resetting inside `CreateWeaponBox`; ammo was not getting reset if `packAmmo` = true, no matter if weapon is not considered "exhaustible"
- Fix: `CanDrop` used inside `PackDeadPlayerItems` in case modders want to make a weapon undroppable.
- Changed usage of constants meaning: GR_PLR_DROP_GUN_* (inside `DeadPlayerWeapons`)

Related to/fixes #520 #620 #758 #821 
Tested, work as intended

PS: Cvar's names/description corrections are welcome

### CODE REVIEW

**Comment by StevenKal** [comment]
Nice! But maybe dropping "all weapons" when "GR_PLR_DROP_GUN_ALL" is set could be added (all primary & secondary).
And keep "GR_PLR_DROP_GUN_ACTIVE" for only active weapon on death (if, it is a primary or secondary, however, get & drop best weapon). But a new rule like "GR_PLR_DROP_GUN_BEST" could also be welcomed.
Or, make "GR_PLR_DROP_GUN_ACTIVE" drops best weapon (default behavior), but this does not really match to the rule name...
So, questionable.

File: regamedll/dlls/multiplay_gamerules.cpp [hunk 1]
```diff
@@ -4332,11 +4332,27 @@ LINK_HOOK_CLASS_CUSTOM_CHAIN(int, CHalfLifeMultiplay, CSGameRules, DeadPlayerWea
 
 int EXT_FUNC CHalfLifeMultiplay::__API_HOOK(DeadPlayerWeapons)(CBasePlayer *pPlayer)
 {
+#ifdef REGAMEDLL_ADD
+	switch((int)weapondrop.value)
```
Comment (wopox1337) [inline regamedll/dlls/multiplay_gamerules.cpp:4336]:
```suggestion
	switch ((int)weapondrop.value)
```

File: regamedll/dlls/multiplay_gamerules.cpp [hunk 2]
```diff
@@ -4332,11 +4332,27 @@ LINK_HOOK_CLASS_CUSTOM_CHAIN(int, CHalfLifeMultiplay, CSGameRules, DeadPlayerWea
...
+			return GR_PLR_DROP_GUN_NO;
+	}
+#endif
 	return GR_PLR_DROP_GUN_ACTIVE;
 }
 
 int CHalfLifeMultiplay::DeadPlayerAmmo(CBasePlayer *pPlayer)
 {
+#ifdef REGAMEDLL_ADD
+	if(ammodrop.value == 0.0f)
```
Comment (wopox1337) [inline regamedll/dlls/multiplay_gamerules.cpp:4352]:
```suggestion
	if (ammodrop.value == 0.0f)
```

File: regamedll/dlls/player.cpp [hunk 3]
```diff
@@ -1386,78 +1393,106 @@ void PackPlayerNade(CBasePlayer *pPlayer, CBasePlayerItem *pItem, bool packAmmo)
...
 		CBasePlayerItem *pBestItem = nullptr;
 
-		for (int n = 0; n < MAX_ITEM_TYPES; n++)
+#ifdef REGAMEDLL_ADD 
+		if(iPackGun == GR_PLR_DROP_GUN_ACTIVE) 
 		{
-			// there's a weapon here. Should I pack it?
-			CBasePlayerItem *pPlayerItem = m_rgpPlayerItems[n];
+			// check if we've just already dropped our active gun 
+			if(!bSkipPrimSec && m_pActiveItem && m_pActiveItem->CanDrop() && m_pActiveItem->iItemSlot() < KNIFE_SLOT)
```
Comment (wopox1337) [inline regamedll/dlls/player.cpp:1415]:
```suggestion
			if (!bSkipPrimSec && m_pActiveItem && m_pActiveItem->CanDrop() && m_pActiveItem->iItemSlot() < KNIFE_SLOT)
```

File: regamedll/dlls/player.cpp [hunk 4]
```diff
@@ -1386,78 +1393,106 @@ void PackPlayerNade(CBasePlayer *pPlayer, CBasePlayerItem *pItem, bool packAmmo)
...
-			bShieldDropped = true;
+			bSkipPrimSec = true;
 		}
 
 		int nBestWeight = 0;
 		CBasePlayerItem *pBestItem = nullptr;
 
-		for (int n = 0; n < MAX_ITEM_TYPES; n++)
+#ifdef REGAMEDLL_ADD 
+		if(iPackGun == GR_PLR_DROP_GUN_ACTIVE)
```
Comment (wopox1337) [inline regamedll/dlls/player.cpp:1412]:
```suggestion
		if (iPackGun == GR_PLR_DROP_GUN_ACTIVE) 
```

File: regamedll/dlls/player.cpp [hunk 5]
```diff
@@ -1386,78 +1393,106 @@ void PackPlayerNade(CBasePlayer *pPlayer, CBasePlayerItem *pItem, bool packAmmo)
...
+			if(!bSkipPrimSec && m_pActiveItem && m_pActiveItem->CanDrop() && m_pActiveItem->iItemSlot() < KNIFE_SLOT)
+			{
+				pBestItem = m_pActiveItem;
 
-			while (pPlayerItem)
+				// if active item is undroppable, then nothing is dropped
+			}
+
+			// are we allowing nade drop?
+			if((int)nadedrops.value >= 1)
```
Comment (wopox1337) [inline regamedll/dlls/player.cpp:1423]:
```suggestion
			if ((int)nadedrops.value >= 1) 
```

File: regamedll/dlls/player.cpp [hunk 6]
```diff
@@ -1386,78 +1393,106 @@ void PackPlayerNade(CBasePlayer *pPlayer, CBasePlayerItem *pItem, bool packAmmo)
...
+			// are we allowing nade drop?
+			if((int)nadedrops.value >= 1) 
+			{
+				// goto item loop but skip guns
+				iPackGun = GR_PLR_DROP_GUN_ALL;
+				bSkipPrimSec = true;
+			}
+		}
+
+		if(iPackGun == GR_PLR_DROP_GUN_ALL)
```
Comment (wopox1337) [inline regamedll/dlls/player.cpp:1431]:
```suggestion
		if (iPackGun == GR_PLR_DROP_GUN_ALL)
```

**Review by wopox1337 — CHANGES_REQUESTED** [review]
Code style fixes

File: regamedll/dlls/player.cpp [hunk 7]
```diff
@@ -8019,7 +8054,11 @@ CBaseEntity *EXT_FUNC CBasePlayer::__API_HOOK(DropPlayerItem)(const char *pszIte
 			vecAngles,
 			vecVelocity,
 			CGameRules::GetItemKillDelay(),
-			false);
+#ifdef REGAMEDLL_ADD
+			(int)ammodrop.value >= 2 ? true : 
+#endif
+			false
```
Comment (wopox1337) [inline regamedll/dlls/player.cpp:8060]: Poor readability. Need to put in a separate variable `bool packAmmo`.

**Review by wopox1337 — CHANGES_REQUESTED** [review]

File: regamedll/dlls/player.cpp (hunk 7 above)
Comment (dystopm) [reply regamedll/dlls/player.cpp:8060]: I didn't create an extra variable to hold that value because I'm stick to the fact we should keep the same behaviour of the original game but I'm concerned of how memory allocation impacts that restriction imposed in this project, maybe I'm too literal. So I guess there's no problem with that. I appreciate your observations and time, will do it now

Comment (wopox1337) [reply regamedll/dlls/player.cpp:8060]: @dystopm 
>  but I'm concerned of how memory allocation impacts that restriction imposed in this project

Could you please elaborate on that, maybe I misunderstood you?

Comment (dystopm) [reply regamedll/dlls/player.cpp]: > @dystopm
> 
> > but I'm concerned of how memory allocation impacts that restriction imposed in this project
> 
> Could you please elaborate on that, maybe I misunderstood you?

Don't worry, I use to be too literal. I've been following this project and Arkshine's first attempt to reverse engineer CS library, using IDA Pro to play with memory addresses and all that kind of things, there's still a lot of things that I need to learn before doubting 😝 ; changes were commited also

File: regamedll/dlls/player.cpp [hunk 8]
```diff
@@ -8013,13 +8048,20 @@ CBaseEntity *EXT_FUNC CBasePlayer::__API_HOOK(DropPlayerItem)(const char *pszIte
 		Vector vecAngles   = pev->angles;
 		Vector vecVelocity = gpGlobals->v_forward * 300 + gpGlobals->v_forward * 100;
 
+#ifdef REGAMEDLL_ADD
+		bool bPackAmmo = (int)ammodrop.value >= 2 ? true : false;
+#else
+		bool bPackAmmo = false;
+#endif
```
Comment (wopox1337) [inline regamedll/dlls/player.cpp:8055]:
```suggestion
bool bPackAmmo = false;

#ifdef REGAMEDLL_ADD
    if (ammodrop.value >= 2.0f)
		    bPackAmmo = true;
#else
```

It might be more comfortable to read this way.

**Comment by dystopm** [comment]
> Nice! But maybe dropping "all weapons" when "GR_PLR_DROP_GUN_ALL" is set could be added (all primary & secondary). And keep "GR_PLR_DROP_GUN_ACTIVE" for only active weapon on death (if, it is a primary or secondary, however, get & drop best weapon). But a new rule like "GR_PLR_DROP_GUN_BEST" could also be welcomed. Or, make "GR_PLR_DROP_GUN_ACTIVE" drops best primary/secondary weapon (default behavior), but this does not really match to the rule name... So, questionable.

I've been thinking on a nice way to implement that. Thing is: drop weapons separately guided by player angles (left to right, depending on how many weapons player will drop) or drop weapons in forward, each one further forward than the last (easy implementation, just store them into an array and apply an scalar to their current velocity since all are going forward) - will take a look and then test

**Comment by dystopm** [comment]
Added new behaviour: mp_weapondrop 3 will ensure players drop all their weapons (primary and secondary) after death. It basically uses the same pattern but every box created has more velocity than its predecessor. Depending on physics, players can stand still, be on mid air or just running fast, and that will affect weapons drop spreading.

hl_g8EpmJvJFR
hl_PXYqh0CfDC
hl_taBKuYQ8LT

Used sv_cheats 1 and impulse 255 for testing purposes only, because in legit scenarios players can have at least 2 weapons.

Credits to @metita for his help in test scenarios.

File: regamedll/dlls/multiplay_gamerules.cpp [hunk 9]
```diff
@@ -4332,11 +4332,29 @@ LINK_HOOK_CLASS_CUSTOM_CHAIN(int, CHalfLifeMultiplay, CSGameRules, DeadPlayerWea
...
-	return GR_PLR_DROP_GUN_ACTIVE;
+#ifdef REGAMEDLL_ADD
+	switch ((int)weapondrop.value)
+	{
+		case 3:
+			return GR_PLR_DROP_GUN_ALL;
+		case 2: 
+			break;
+		case 1:
+			return GR_PLR_DROP_GUN_BEST;
```
Comment (wopox1337) [inline regamedll/dlls/multiplay_gamerules.cpp:4343]: @dystopm `mp_weapondrop 1` now changes default behavior.

Will drop `GR_PLR_DROP_GUN_BEST` instead `GR_PLR_DROP_GUN_ACTIVE`

**Review by wopox1337 — CHANGES_REQUESTED** [review]

File: regamedll/dlls/multiplay_gamerules.cpp (hunk 9 above)
Comment (dystopm) [reply regamedll/dlls/multiplay_gamerules.cpp:4343]: @wopox1337 I need your opinion here.

If you take a deep look, GR_PLR_DROP_GUN_ACTIVE value is never used, it's just a placeholder return value because inside `PackDeadPlayerItems`, the return value asked for `DeadPlayerWeapons` to drop player guns is `!= GR_PLR_DROP_GUN_NO` for dropping the best weapon on player -> which has nothing to do with the meaning of the constant. I've just corrected that, but yes and no, mp_weapondrop 1 changes the internal return value but not the default behaviour 😬 and code inside `PackDeadPlayerItems` acts normally on fixed and original behaviour

Comment (dystopm) [reply regamedll/dlls/multiplay_gamerules.cpp:4343]: @wopox1337 original CS code:

notepad++_vy5teHYUY6

with 9 = GR_PLR_DROP_GUN_NO

Comment (dystopm) [reply regamedll/dlls/multiplay_gamerules.cpp:4343]: @wopox1337 so it's a yes?

Comment (wopox1337) [reply regamedll/dlls/multiplay_gamerules.cpp:4343]: Sure

**Review by wopox1337 — APPROVED** [review]

File: regamedll/dlls/player.cpp [hunk 10]
```diff
@@ -1386,78 +1395,130 @@ void PackPlayerNade(CBasePlayer *pPlayer, CBasePlayerItem *pItem, bool packAmmo)
 void CBasePlayer::PackDeadPlayerItems()
```
Comment (wopox1337) [inline regamedll/dlls/player.cpp:1395]: The CBasePlayer::PackDeadPlayerItems() function looks terrible (it's not because of your PR). It needs refactoring (later), for the logic is hard to follow. 😞

**Review by wopox1337 — APPROVED** [review]

File: regamedll/dlls/player.cpp (hunk 10 above)
Comment (dystopm) [reply regamedll/dlls/player.cpp:1395]: @wopox1337 yeah you're right, I just tried to stick to the original code but indeed it is certainly bad coded
